File: 0525-contiguous-array/0525-contiguous-array.rs
use std::collections::HashMap;

impl Solution {
    pub fn find_max_length(nums: Vec<i32>) -> i32 {
        let (n, mut sum, mut res) = (nums.len(), 0, 0);
        let mut mp: HashMap<i32, i32> = HashMap::new();
        mp.insert(0, -1);
        for i in 0..n {
            if nums[i] == 1 {
                sum += 1;
            } else {
                sum -= 1;
            }
            if let Some(&idx) = mp.get(&sum) {
                res = res.max(i as i32 - idx);
            } else {
                mp.insert(sum, i as i32);
            }
        }
        res
    }
}

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size(), sum = 0, res = 0;
        unordered_map<int, int> mp;
        mp[0] = -1;
        for (int i = 0; i < n; i++) {
            sum += nums[i] == 1 ? 1 : -1;
            if (mp.find(sum) != mp.end()) {
                res = max(res, i - mp[sum]);
            } else {
                mp[sum] = i;
            }
        }
        return res;
    }
};